docs(blog): Add localised strings Kentico cloud to Gatsby and GraphQL post (#16601)
gatsbyjs/gatsby: docs/blog/2019-08-13-Localised-strings-from-Kentico-Cloud-GatsbyJS-GraphQL/index.md

---

> title: Adding and retrieving localised strings from Kentico Cloud to GatsbyJS and GraphQL

Reviewer: To match the [style guide](https://www.gatsbyjs.org/contributing/gatsby-style-guide/#prefer-us-english) it seems like we should probably opt for the US style "localized" instead:

```suggestion
title: Adding and retrieving localized strings from Kentico Cloud to GatsbyJS and GraphQL
```

Author: Thanks I will change this

date: 2019-08-13

> author: "Ilesh Mistry"

Reviewer: Normally it doesn't matter whether or not we use `"`'s in frontmatter, but in this case we match the author with the info in the `author.yaml` file and with quotes (when I run the site locally) it fails to find the correct author for your name:

```suggestion
author: Ilesh Mistry
```

image: "images/country-clocks.jpg"
imageTitle: Different country flag clocks on the wall showing the different times 
excerpt: "This blog post talks about how to add and retrieve localised strings from Kentico Cloud by using GatsbyJS and GraphQL"
canonicalLink: https://www.ileshmistry.com/adding-and-retrieving-localised-strings-from-kentico-cloud-to-gatsbyjs-and-graphql/
publishedAt: Ilesh Mistry's website
tags: ["gatsby", "graphql", "kentico-cloud", "kentico"]
---

Every website or application tends to have text elements which are typically static and rarely change – think of the items in your header or footer (e.g. copyright text). For many sites, the temptation is to simply hard code this into the master template or templates. However, when you throw multilingual websites or shared templates into the mix, this swiftly becomes a bad idea. Handing the editors the ability to content manage these items becomes a necessity. 

> If you're familiar with content management systems like [Kentico EMS](https://www.kentico.com/), then you will have come across the term [Resource Strings or Localised Strings](https://docs.kentico.com/k12sp/multilingual-websites/setting-up-a-multilingual-user-interface/working-with-resource-strings). These localised strings replace the static text across your site and are a combination of a 'key' and 'value' pairing. The CMS consumes the 'key' to retrieve the 'value' of your chosen culture or language as and when required. The use of these localised strings gives editors the additional content management control required.

Reviewer: ```suggestion
If you're familiar with content management systems like [Kentico EMS](https://www.kentico.com/), then you will have come across the term [Resource Strings or Localized Strings](https://docs.kentico.com/k12sp/multilingual-websites/setting-up-a-multilingual-user-interface/working-with-resource-strings). These localized strings replace the static text across your site and are a combination of a 'key' and 'value' pairing. The CMS consumes the 'key' to retrieve the 'value' of your chosen culture or language as and when required. The use of these localized strings gives editors the additional content management control required.
```

> Most CMS offer methods for handling these localised strings. However, with “headless”, API-first CMS, there’s no out-of-the-box feature and a best practice has to be established. I'm going to use [Kentico Cloud](https://kenticocloud.com/), a CaaS ([Content as a Service](https://en.wikipedia.org/wiki/Content_as_a_service)) platform which focuses on a content-first approach, providing you the ability to create, manage and distribute content without worrying about how and where it will be displayed. If you need to know more about Kentico Cloud and how to get started with it, you can catch up on the product through my previous post, [Getting started with Kentico Cloud](https://www.ileshmistry.com/getting-started-with-kentico-cloud/).

Reviewer: ```suggestion
Most CMSs offer methods for handling these localized strings. However, with a “headless”, API-first CMS, there’s no out-of-the-box feature, and a best practice has to be established. I'm going to use [Kentico Cloud](https://kenticocloud.com/), a CaaS ([Content as a Service](https://en.wikipedia.org/wiki/Content_as_a_service)) platform which focuses on a content-first approach, providing you the ability to create, manage, and distribute content without worrying about how and where it will be displayed. If you need to know more about Kentico Cloud and how to get started with it, you can catch up on the product through my previous post, [Getting started with Kentico Cloud](https://www.ileshmistry.com/getting-started-with-kentico-cloud/).
```

*Disclaimer – Languages only apply to certain Kentico Cloud plans. For the approach in this article, I’m assuming cultures have been enabled in your selected subscription plan!*

> The flexibility of Kentico Cloud means there are many ways to achieve the same goal. Having been through this several times, I’m going to step you through a best practice approach to creating and retrieving localised strings.

Reviewer: ```suggestion
The flexibility of Kentico Cloud means there are many ways to achieve the same goal. Having been through this several times, I’m going to step you through a best practice approach to creating and retrieving localized strings.
```

I'd keep suggesting changes for these but it's probably easier for you just to do a big find/replace 🙂

Whichever route you’ve taken, once you know the culture code then you can get any ‘key’ and ‘value’ pairs for the required culture.

## Setting up your content type
If you want to look at building localised strings, the first step is to head to your [Content Types](https://docs.kenticocloud.com/tutorials/set-up-projects/define-content-models/creating-and-deleting-content-types). You would create a Localisation String content item.

Here is a [Content Item](https://docs.kenticocloud.com/tutorials/write-and-collaborate/write-content/adding-content-items) derived from a '*Localised String*' Content Type, which has a single field for the 'value'.

![An example of the localised content item](images/Localise-Content-Item.png)

As I mentioned earlier, localised strings work in a pair of 'key' and 'value'. The good thing about Kentico Cloud is that the 'key' is something you could derive from the 'Codename' for the content item and it saves you creating additional fields for it.

![An example of the localised content item, showing the codename](images/Localise-Content-Item-Codename.png)

> You will most likely be utilising the IDs for the content item, but it's useful to have the codename as well.

Reviewer: ```suggestion
You will most likely be utilizing the IDs for the content item, but it's useful to have the codename as well.
```

![A view of the GraphiQL preview of retrieving language nodes](images/GraphiQL-retrieve-lang-nodes.png)

You can retrieve all localisation strings using the following GraphQL query.
![Retrieving all language versions of the localised strings](images/GraphiQL-retrieve-lang-variants.png)

However, to retrieve the actual 'value', there’s a little more to do in your GraphQL query.

Here is an example of a GraphQL query where I am using the CodeName.
![GraphiQL retrieve language nodes localised strings condition codename](images/GraphiQL-retrieve-lang-variants-based-on-condition-codename.png)

And, here is an example of a GraphQL query where I am using the ID.
![GraphiQL retrieve language nodes localised strings condition ID](images/GraphiQL-retrieve-lang-variants-based-on-condition-id.png)

As you can see, once you know which 'key' you need, then it's pretty simple to get the 'value' and also detect the language variant you need to retrieve in GatsbyJS using this GraphQL query format.

## Separate projects

> Before we go any further, it is important to take stock of your localisation string content items. Creating culture versions of each item doesn’t increase your number of Content Items so you’re unlikely to be jumping up through plans. However, the number of localisation string Content Items is something to keep in mind. Also, consider the number of environments/projects you have in your subscription, which could multiple the number of Content Items you have in total.

Reviewer: ```suggestion
Before we go any further, it is important to take stock of your localisation string content items. Creating culture versions of each item doesn’t increase your number of Content Items so you’re unlikely to be jumping up through plans. However, the number of localisation string Content Items is something to keep in mind. Also, consider the number of environments/projects you have in your subscription, which could multiply the number of Content Items you have in total.
```

> We would always advocate pragmatism when creating your content models. In addition, maybe consider creating the localisation strings as a separate project that feeds into all the projects that need to utilise them. This is a solution that has worked across several of our projects and ensures that all localised strings are available to all projects and all environments, removing the overhead for updates.

Reviewer: ```suggestion
We would always advocate pragmatism when creating your content models. In addition, maybe consider creating the localization strings as a separate project that feeds into all the projects that need to utilize them. This is a solution that has worked across several of our projects and ensures that all localized strings are available to all projects and all environments, removing the overhead for updates.
```

Maybe find and replace other places where utilize pops up?

Author: yep, that's the best way and hopefully, I've now done this 👍

> We would always advocate pragmatism when creating your content models. In addition, maybe consider creating the localization strings as a separate project that feeds into all the projects that need to utilise them. This is a solution that has worked across several of our projects and ensures that all localized strings are available to all projects and all environments, removing the overhead for updates.

Reviewer: ```suggestion
We would always advocate pragmatism when creating your content models. In addition, maybe consider creating the localization strings as a separate project that feeds into all the projects that need to utilize them. This is a solution that has worked across several of our projects and ensures that all localized strings are available to all projects and all environments, removing the overhead for updates.
```

## Summary - A recap of the steps involved in creating localized content in Kentico Cloud

There’s a lot to consider there so let's recap:

> * Create a Content Type for localise string items.

Reviewer: ```suggestion
* Create a Content Type for localize string items.
```

* Add appropriate Content Items with the values for each localized string item.
* Create the relevant culture alternatives for the content items created.
* Once you know the desired culture you would then look in GraphQL nodes for the localized strings you need to show and use the filter option to help you bring specific localized strings back.

> * Utilise the query within your GatsbyJS layout/React code.

Reviewer: ```suggestion
* Utilize the query within your GatsbyJS layout/React code.
```

Author: @gillkyle thanks for this. Hopefully, I've amended the few that were missed. 

I've also run prettier and this should also come through.
